docs/api_requests/block-vhost-user.md

```diff
@@ -0,0 +1,189 @@
+the backend:
+
+1. Device initialisation. When a vhost-user device is created, Firecracker
+   connects to the corresponding UDS socket and negotiates Virtio and Vhost
+   features with backend and retrieves device configuration.
+1. Device activation. When the guest driver finishes setting up the device,
+   Firecracker shares memory tables and Virtio queue information with the backend.
+   As a part of this, Firecracker shares file descriptors for guest's memory
+   regions, as well as file descriptors for queue notifications.
+
+## Advantages
+
+While vhost-user block is considered an optimisation to Firecracker IO, a naive
+implementation of the backend is not going to improve performance.
+
+The major advantage of using a vhost-user device is the backend can implement
+custom processing logic. It can use intelligent algorithms to serve block
+requests, eg by fetching the block device data over the network or using
+encryption or using sophisticated readahead logic. In such cases,
+the performance improvement will be coming from the fact that the custom logic
+is implemented in the same process that handles Virtio queues, which reduces
+the number of required context switches.
+
+## Backends
+
+There are a number of open source implementations of a vhost-user backend
+available for reference:
+
+1. [Qemu backend](https://github.com/qemu/qemu/tree/master/contrib/vhost-user-blk)
+1. [Cloud Hypervisor backend](https://github.com/cloud-hypervisor/cloud-hypervisor/tree/main/vhost_user_block)
+1. [crosvm backend](https://github.com/google/crosvm/blob/main/devices/src/virtio/vhost/user/device/block.rs)
+1. [SPDK backend](https://github.com/spdk/spdk/blob/master/lib/vhost/vhost_blk.c)
+
+## Security considerations
+
+### Guest memory sharing
+
+By design, a vhost-user frontend must share file descriptors
+of all guest memory regions to the backend. In order to achive that,
+guest memory is created as a [memfd](https://man7.org/linux/man-pages/man2/memfd_create.2.html)
+and mapped as `MAP_SHARED`.
+
+#### File descriptor in procfs
+
+An open `memfd` is reflected in `procfs` as any other open file descriptor:
+
+```shell
+$ ls -l /proc/{pid}/fd | grep memfd
+lrwx------ 1 1234 1234 64 Nov  2 13:39 32 -> /memfd:guest_mem (deleted)
+```
+
+Any process on the host that has access to this file in `procfs` will be able
+to map the file descriptor and observe runtime behaviour of the guest.
+
+At the moment, Firecracker does not close the `memfd`, because it must remain open
+until all the configured vhost-user devices have been activated and their info
+shared with the backends. This kind of tracking is not implemented in Firecracker,
+but may be implemented in the future. Meanwhile, users need to make sure that
+the access to the Firecracker's `procfs` tree is restricted to trusted processes
+on the host.
+
+On the backend side, it is advised that the backend closes
+the guest memory region file descriptors after mapping them into the address
+space.
```

Comment:
What effect does that have in the `memfd` file appearing in procfs? Also:

`mapping them into the address space.` -> `mapping them into its own address space`.

Reply:
comments on this?

Reply:
>What effect does that have in the memfd file appearing in procfs?

Does the following not answer your question?

>Any process on the host that has access to this file in `procfs` will be able
to map the file descriptor and observe runtime behaviour of the guest.

>mapping them into the address space. -> mapping them into its own address space.

I don't think it is essential, because it doesn't look possible to map something to another process's address space, does it?

Reply:
>mapping them into the address space. -> mapping them into its own address space.

added, since I was at it anyway

Reply:
> Does the following not answer your question?

    Any process on the host that has access to this file in procfs will be able
    to map the file descriptor and observe runtime behaviour of the guest.

It does! Sorry, I missed that.